token.go

```diff
@@ -178,3 +179,29 @@ func (m TokenModule) postTokenHandler(c *gin.Context) {
 
 	c.JSON(http.StatusCreated, token)
 }
+
+// postTokenHandler godoc
+// @summary Put token in database.
+// @description Put token in database. Provider in post object has to exists or should be empty.
+// @description Provider will has to be updated token ID during this operation.
+// @tags token
+// @accept json
+// @produce json
+// @param token body TokenWithProviderID _ "token object"
+// @success 201 {object} Token
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 401 {object} problem.Response "Unauthorized or missing jwt token"
+// @failure 404 {object} problem.Response "Referenced provider not found"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /token [put]
+func (m TokenModule) putTokenHandler(c *gin.Context) {
+	var token Token
+	err := c.ShouldBindJSON(&token)
+	if err != nil {
+		problem.WriteInvalidBindError(c, err, "One or more parameters failed to parse when reading the request body.")
+		return
+	}
+	var placedToken Token
+	m.Database.Where(token).FirstOrCreate(&placedToken)
```

Comment:
Have not tackled this issue that CodeQL is warning about in this repo yet. But it seems it wants us to do something like this instead:

```suggestion
	m.Database.
		Where(&token, tokenFieldToken, tokenFieldUserName).
		FirstOrCreate(&placedToken)
```

Based on recommendation from https://gorm.io/docs/security.html

Not sure though, as the security warning only shows recommendations on regular `database/sql` usage, and the GORM security page only shows info on using SQL column names directly like `.Where("token = ?", token.Token)`. Want to refrain from using column names in strings, and `.Where(tokenFieldToken + "= ?", token.Token)` is just overly verbose

https://github.com/iver-wharf/wharf-api/security/code-scanning/14?query=ref%3Arefs%2Fpull%2F26%2Fmerge

Reply:
As far as my tests and understanding on this goes-
The sql injection mentioned in the gorm security doc you linked is only available when gorm isnt given the opportinuty to split the collumn name from the value its being set to. 

This is not relevant in our case because the first object (int this case token) is adequately constructed. The named values are split by gorm. In this case in a key:value pairing.

The resulting sql query from both with/without the sugestion is as follows:
```bash
[10.150ms] [rows:1] INSERT INTO "token" ("token","user_name") VALUES ('sample-token\'; drop table branch;','') RETURNING "token_id"
[GIN] 2021/06/02 - 17:12:08 | 202 |         3m54s |      172.19.0.1 | PUT      "/api/token"
[GIN] 2021/06/02 - 17:12:08 | 400 |            0s |      172.19.0.1 | POST
```
These results are created with a post to /import/github with a body containing 
```json
{
    "group": "iver-wharf",
    "project": "wharf-provider-github",
    "token": "sample-token'; drop table branch;",
    "url": "https://api.github.com"
}
```
As you can see the escape is correct in both cases.

This is also works (and escapes) correctly if the code is changed to something like 
```go
m.Database.Where("Token=?", token.Token)
```


The issue of sql injection first seems to arise when this is changed to
```go
m.Database.Where("token=" + token.Token)
```
Even in this scenario it still uses PDOs so it appears multiple sql statements are impossible but it seems we could go fishing for other friendly tokens- (tested and works)
```json
{
    "group": "iver-wharf",
    "project": "wharf-provider-github",
    "token": "'sample-token' OR token not like 'sample-token'",
    "url": "https://token.fishing.me"
}
```

Final words - If i didnt miss anything then i propse we avoid code bloat by not passing unessacery function params. If i did miss something about how this works i would love to hear more.

Reply:
Nice research! I have never tested this so it's really nice to know for sure.

I've marked the CodeQL as "False positive" in https://github.com/iver-wharf/wharf-api/security/code-scanning/15?query=ref%3Arefs%2Fpull%2F26%2Fmerge